Migrate `Transport` to use pydantic for configuration

sphuber opened the pull request:
Following #6255, we introduce here pydantic models also to the Transport module for use in configuring computers.

sphuber committed:
WIP: Migrate `Transport` to use pydantic for configuration

agoscinski commented:
Is it worth to tag this with a v3.0.0 milestone where we could introduce such breaking changes?

sphuber commented:
> Is it worth to tag this with a v3.0.0 milestone where we could introduce such breaking changes?

The breaking changes are actually quite minimal. I don't remember by heart, but it may be just the order in which options are prompted for and some other minimal things. Mostly the changes are just in implementation.

And I am not sure if there actually ever will be a v3 😅 There are no concrete plans for now in any case. But definitely worth a discussion at some point perhaps.

edan-bainglass commented:
@agoscinski @khsrali working on this one. Rebased on `main`. Need to add `Model`s to the new `Transport` classes (blocking, async). Then tests. Then we should be good. Let's see.

edan-bainglass committed:
Update `SshTransport.Model` parent to `BlockingTransport.Model` for consistency

There are no additional fields. This is just a reference to the parent `Transport.Model`. Still best to follow the hierarchical chain for consistency.

edan-bainglass commented:
@agoscinski do we want to add model testing for transport similar to that of the orm module?

edan-bainglass committed:
Make localhost configure non-interactive during CI environment setup

edan-bainglass commented:
@agoscinski maybe check that last commit regarding the non-interactive option for the localhost config. I guess this wasn't an issue before due to loose code but now is due to Seb's updates. We use `--non-interactive` when configuring the slurm-ssh computer, so this change seemed consistent.

khsrali commented:
> @agoscinski @khsrali working on this one. Rebased on `main`. Possibly need to add `Model`s to the new blocking/async `Transport` classes (or maybe not? I don't see any new fields - @khsrali?). Then tests. Then we should be good. Let's see.

https://github.com/aiidateam/aiida-core/blob/71fc14f3c2a501ff8d704d20df76a297edc8e8bc/src/aiida/transports/plugins/ssh_async.py#L70C1-L109C1

edan-bainglass committed:
Add `AsyncSshTransport.Model`

edan-bainglass commented:
@khsrali can you give a look to the new `AsyncSshTransport.Model`?

edan-bainglass committed:
Lift `_DEFAULT_max_io_allowed` to module level

edan-bainglass commented:
@khsrali please review 🙏

khsrali commented:
I apologize @edan-bainglass 

This was a super busy week.. now I'm out on a vacation, and don't have my laptop with me..

It's quite painful to review with my phone 🥲

Will be back to work on Tuesday the 20th. 
I hope that's not too late. If you urgently need this, maybe another team member can review?

P.S. also Alex is on vacation, so I think this can wait from the *release* point of view.

khsrali commented on src/aiida/transports/plugins/ssh_async.py:
```diff
@@ -108,6 +110,28 @@ class AsyncSshTransport(AsyncTransport):
         ),
     ]
 
+    class Model(AsyncTransport.Model):
```
Why are the previous stuff are also kept? 
Line 75-110

edan-bainglass replied:
As discussed, this is unrelated. The "previous stuff" is used in the CLI. However, still to be discussed is if that should remain, or if we use the model. If possible, good if we don't mess with the current UX, which informs the user of issues when they occur. The use of pydantic models will likely defer notification of issues until the end of the CLI process. Not ideal or friendly.

khsrali commented on src/aiida/transports/transport.py:
```diff
@@ -97,6 +100,21 @@ class Transport(abc.ABC):
         ),
     ]
 
+    class Model(BaseModel):
```
The same question here

khsrali commented on .github/workflows/setup.sh:
```diff
@@ -15,7 +15,7 @@ verdi setup --non-interactive --config "${CONFIG}/profile.yaml"
 # set up localhost computer
 verdi computer setup --non-interactive --config "${CONFIG}/localhost.yaml"
-verdi computer configure core.local localhost --config "${CONFIG}/localhost-config.yaml"
+verdi computer configure core.local localhost --non-interactive --config "${CONFIG}/localhost-config.yaml"
```
I wonder why this was not needed before

khsrali replied:
Update: 
@edan-bainglass reports he cannot configure a computer even with this change

khsrali replied:
That could also potentially mean that, the tests we have in place are not adequate to capture that failure.

@edan-bainglass can you please tailor the error you get?
Thanks!

khsrali reviewed (commented):
I haven't check the changes in cmd_computer, carefully. 
I have to check if we have complete test suits for that.

edan-bainglass commented on src/aiida/cmdline/commands/cmd_computer.py:
```diff
@@ -659,25 +659,40 @@ def _dry_run_callback(pks):
     echo.echo_success(f'Computer `{label}` {"and all its associated nodes " if associated_nodes_pk else ""}deleted.')
 
 
-class LazyConfigureGroup(VerdiCommandGroup):
```
Part of @sphuber's original work. It seems now more standard w.r.t the way the code command is handled, which was already using pydantic I think since the abstract code refactoring.

edan-bainglass replied:
I'm not sure why this is the case, but re-instating the `LazyConfigureGroup` class breaks the `TestVerdiComputerConfigure.test_ssh_interactive` test. The values passed to the CLI command are somehow out of order, leading to the following nonsense:
```
Report: enter ? for help.
Report: enter ! to ignore the default and set no value.
User name [edanb]: 
Port number [22]: 
Look for keys [Y/n]: some_remote_user   <- ?
Error: invalid input
Look for keys [Y/n]: 345
Error: invalid input
Look for keys [Y/n]: no
SSH key file []: 
Connection timeout in s [60]: 
Allow ssh agent [Y/n]: 
SSH proxy jump []: 
SSH proxy command []: 
Compress file transfers [Y/n]: 
GSS auth [False]: 
GSS kex [False]: 
GSS deleg_creds [False]: 
GSS host [localhost]: 
Load system host keys [Y/n]: 
Key policy (RejectPolicy, WarningPolicy, AutoAddPolicy) [RejectPolicy]: 
Use login shell when executing command [Y/n]: 
Connection cooldown time (s) [30.0]: 
Report: Configuring computer test_ssh_interactive for user test@localhost.
Success: test_ssh_interactive successfully configured for test@localhost
```
instead of the correct
```
Report: enter ? for help.
Report: enter ! to ignore the default and set no value.
Use login shell when executing command [Y/n]: 
Connection cooldown time (s) [30.0]: 
User name [edanb]: some_remote_user
Port number [22]: 345
Look for keys [Y/n]: no
SSH key file []: 
Connection timeout (s) [60]: 
Allow ssh agent [y/N]: 
SSH proxy jump []: 
SSH proxy command []: 
Compress file transfers [Y/n]: 
GSS auth [y/N]: 
GSS kex [y/N]: 
GSS deleg_creds [y/N]: 
GSS host []: 
Load system host keys [Y/n]: 
Key policy (RejectPolicy, WarningPolicy, AutoAddPolicy) [RejectPolicy]: 
Report: Configuring computer test_ssh_interactive for user test@localhost.
Success: test_ssh_interactive successfully configured for test@localhost
```
At the moment, it's not clear what broke this part. I suppose @sphuber encountered this and fixed it. Unfortunately (uncharacteristically) not documented 😞

edan-bainglass commented:
@sphuber I pinged you on at least one comment regarding the original work. However, it would be great if you could provide a description to this PR briefly detailing the motivation and summarizing the changes.

edan-bainglass committed:
Fix `AsyncSshTransport.Model`

edan-bainglass commented:
@khsrali @agoscinski we need to revisit the `test_ssh_interactive` test. It is failing, but in practice, configuring a computer interactively works. Perhaps a bad test. I'm looking into it.

agoscinski commented:
@khsrali can you look at this, I have the feeling it is a related bug as you fixed it in https://github.com/aiidateam/aiida-core/pull/6919. It looks from a quick glance that the CLI option are now created differentl only by defining a pydantic model https://github.com/aiidateam/aiida-core/blob/6b85f047cf3e9e340a6dc4ac25aba522c5dfd038/src/aiida/cmdline/groups/dynamic.py#L158.

edan-bainglass committed:
Discard `non-interactive` positional argument in `configure_computer`

edan-bainglass committed:
Revert unrelated changes to the CLI

edan-bainglass commented:
@agoscinski @khsrali I think perhaps we were wrong to remove the changes to the CLI. Unlike the ORM pydantic models introduced in #6842, this PR introduces pydantic models to the Transport classes solely for use in the CLI. Not sure if @sphuber expected more of them (e.g., serialization), but in this PR, no such mechanism is introduced. The serialization mechanics introduced in #6842 are limited to Entity subclasses, of which the Transport module is not a part of. So there is no straight forward (de)serialization here.

If the above is accurate, I would close #6924 and reintroduce the CLI changes here, then add a test that the model is operational, whatever that means.

Comments?

khsrali commented:
Transport is not an Entity subclass, and the serialization/deserialization machinery from #6842 only applies to entities. So here the pydantic model does nothing the existing _valid_auth_options doesn't have. 
I see very limited value in this. Does not worth the time.
